---
title: "Evaluation of segmentation"
author: "Wancen Mu"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
output:
  rmarkdown::html_document:
    highlight: tango
vignette: |
  %\VignetteIndexEntry{Evaluation of segmentation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r}
library(EnsDb.Hsapiens.v86)
edb <- EnsDb.Hsapiens.v86
g <- genes(edb, filter = AnnotationFilterList(GeneIdFilter("ENSG", "startsWith")))
library(GenomeInfoDb)
gr <- keepStandardChromosomes(g, pruning.mode = "coarse", species = "Homo_sapiens")
seqlevelsStyle(gr) <- "UCSC"
seqlevels(gr)
gr <- sort(gr)
# g1 <- keepSeqlevels(g, as.character(seq_len(9)), pruning.mode = "coarse")
```

## Import deny region

```{r}
library(nullrangesData)
data("deny")
```

```{r}
library(nullranges)
library(plyranges)
```

# Gene density distribution-original dataset

## CBS segmentation

```{r}
deny2 <- deny %>% filter(width(deny) > 500)
Ls <- 1e6
seg <- segment_density(gr, n = 3, Ls = Ls, deny = deny2, type = "cbs", boxplot = TRUE)
```

# Gene density distribution-segmentated blockbootstrap-CBS

```{r}
# Test on 100 times bootstrap ~17s
# L_b <- 5e5
# system.time(res_wochr <- seg_bootstrap_granges(seg, gr, L_b, R = 100, within_chrom = FALSE, proportion_length = FALSE, coarse = TRUE))[[3]]
# system.time(res_wochr_propl <- seg_bootstrap_granges(seg, gr, L_b, R = 100, within_chrom = FALSE, coarse = TRUE))[[3]]
# system.time(res_wchr <- seg_bootstrap_granges(seg, gr, L_b,R = 100, within_chrom = TRUE, proportion_length = FALSE, coarse = TRUE))[[3]]
# system.time(res_wchr <- seg_bootstrap_granges(seg, gr, L_b,R = 100, within_chrom = TRUE, proportion_length = TRUE, coarse = TRUE))[[3]]
```

# Test performance on difference segmentation and block length

```{r}
# Ls <- c(1e5, 1e6, 5e6)
# L_b <- c(5e4, 5e5, 1e6)
# 
# mapply(function(Ls, L_b) {
#   seg <- segment_density(g1, n = 3, Ls = Ls, type = "CBS", plot_origin = TRUE)
#   seg_length <- seg %>%
#     group_by(state) %>%
#     summarise(Ls = sum(width)) # derive each states length
# 
#   L_c <- seqlengths(g1)["9"]
#   L_s <- seg_length$Ls
#   system.time(res <- seg_bootstrap_iranges(seg, g1, L_c, L_s, L_b))[[3]]
#   res
#   res <- do.call(c, res)
#   res
#   res <- sort(res)
#   res
#   CBS_boot <- GRanges(seqnames = "9", ranges = res, seqlengths = L_c)
#   counts_cbs_boot <- countOverlaps(query, CBS_boot)
#   eps <- rnorm(length(counts_cbs_boot), 0, .2)
#   print(hist(counts_cbs_boot, breaks = 50, xlab = paste(Ls, "with", L_b, "block size")))
#   print(plot(sqrt(counts_cbs_boot) + eps))
# }, Ls, L_b)
```

# across chromosome

```{r}
# library(ggbio)
# ggbio::autoplot(gr)
```

```{r}
# ggbio::autoplot(res_wchr)
```

```{r}
# ggbio::autoplot(res_ochr)
```


# Toy example on evalidating block bootstrap according to segmentation region

```{r}
library(GenomicRanges)
seg <- GRanges("1", IRanges(c(1, 101, 201, 301), c(100, 200, 300, 400), state = c(1:3, 1)))
x <- GRanges("1", IRanges(c(21, 26, 31, 36, 41, 221, 231, 241), width = 4), seqlengths = c("1" = 400))
library(ggbio)
ggbio::autoplot(x) + xlim(1, 400)
res <- seg_bootstrap_granges(seg, x,R=1, L_b = 100)
# ggbio::autoplot(res) + xlim(1, 400)
```

```{r}
library(GenomicRanges)
gr <- GRanges(rep(c("chr1", "chr2", "chr3"), c(5, 10, 3)),
  IRanges(
    start = c(1, 100, 120, 200, 250, 50, 60, 100, 150, 200, 215, 230, 280, 290, 400, 1, 100, 130),
    width = c(20, 5, 5, 30, 20, 30, 30, 40, 15, 15, 20, 5, 5, 5, 20, 30, 80, 60)
  ),
  seqlengths = c(chr1 = 300, chr2 = 450, chr3 = 200)
)
autoplot(gr)
```
