---
title: "Segmented block bootstrap"
author: "Wancen Mu"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
output:
  rmarkdown::html_document:
    highlight: tango
    toc: true
    toc_float: true	
vignette: |
  %\VignetteIndexEntry{Segmented block bootstrap}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction to segmented block bootstrap

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=5, fig.height=5)
```

As proposed by Bickel et al (2010), *nullranges* contains an
implementation of a block bootstrap, such that features are sampled
from the genome in blocks. Blocks are sampled and placed within
regions of a genomic segmentation. That is, for a genome segmented
into states 1,2,...,S, blocks from state $s$ will be used to tile the
ranges of state $s$ in each bootstrap sample.

The segmented block bootstrap has two options, either to:

* Segment the genome using feature density, e.g. gene density
* Use exiting segmentation (e.g. ChromHMM, etc.) downloaded from
  AnnotationHub or external to Bioconductor (BED files imported with
  *rtracklayer*)

In this vignette, we give an example of segmenting the hg38 genome by
Ensembl gene density, then we profile the time to generate a single
block bootstrap sample. Finally, we use a toy dataset to visualize
what a segmented block bootstrap sample looks like with respect to a
genome segmentation. Future versions of this vignette will demonstrate
the functions use within an overlap analysis. See also the unsegmented
block bootstrap vignette in *nullranges*.

# Segmentation by gene density

First we obtain the Ensembl genes for segmenting by gene density.

```{r}
suppressPackageStartupMessages(library(EnsDb.Hsapiens.v86))
edb <- EnsDb.Hsapiens.v86
filt <- AnnotationFilterList(GeneIdFilter("ENSG", "startsWith"))
g <- genes(edb, filter = filt)
```

Some processing to align these features with our deny regions and our
features of interest (DHS, defined below).

```{r}
library(GenomeInfoDb)
g <- keepStandardChromosomes(g, pruning.mode = "coarse")
seqlevels(g, pruning.mode="coarse") <- setdiff(seqlevels(g), "MT")
# seqlevelsStyle(g) <- "UCSC" # host issues with NCBI
seqlevels(g) <- paste0("chr", seqlevels(g))
genome(g) <- "hg38"
g <- sortSeqlevels(g)
g <- sort(g)
table(seqnames(g))
```

## Import deny region

Here we import deny regions (as created by Kundaje et al). 

TODO: here instead use Mikhail's *denyranges* package, once it is available.

```{r}
library(nullrangesData)
data("deny")
table(seqnames(deny))
```

## CBS segmentation

We load *nullranges* and *plyranges* packages:

```{r}
library(nullranges)
library(patchwork)
suppressPackageStartupMessages(library(plyranges))
```

We subset the deny ranges to those which are 500 bp or larger. 

TODO: discuss motivation for this.

We use a CBS segmentation as implemented in *DNAcopy* (cite). 

```{r seg-cbs}
set.seed(5)
deny2 <- deny %>% plyranges::filter(width(deny) >= 500)
L_s <- 1e6
seg <- segmentDensity(g, n = 3, L_s = L_s,
                          deny = deny2, type = "cbs")
plots <- lapply(c("ranges","barplot","boxplot"), function(t) plotSegment(seg,deny2, type = t))
plots[[1]]
plots[[2]] + plots[[3]]
```

Note here, the default *ranges* plot gives whole genome and every fractured bind regions represents state transformations happens. However, some transformations within small ranges cannot be visualized, e.g 1kb. If user want to look into specific ranges of segmentation state, the *region* argument is flexible to support. 

```{r}
region=GRanges("chr16",IRanges(3e7,4e7))
plotSegment(seg,deny2,type="ranges",region=region)
```


## Alternatively: HMM segmentation

Here we use an alternative segmentation implemented in *RcppHMM* (cite).

```{r seg-hmm}
seg_hmm <- segmentDensity(g, n = 3, L_s = L_s,
                          deny = deny2, type = "hmm")
plots <- lapply(c("ranges","barplot","boxplot"), function(t) plotSegment(seg_hmm,deny2, type = t))
plots[[1]]
plots[[2]] + plots[[3]]
```

# Segmented block bootstrap within chromosome

# Timing on DHS peaks

Here, we use a set of DNase hypersensitivity sites (DHS) from the
ENCODE project in A549 cell line. Here, for speed, we load an existing
data object which has been processed with the following steps:

* Download DNase hypersensitivity peaks in A549 from *AnnotationHub* 
* Subset to standard chromosomes and remove mitochondria (more work
needed to handle these cases internally).
* Use a chain file from UCSC to lift ranges from hg19 to hg38
* Sort the DHS features to be bootstrapped

```{r}
data("DHSA549Hg38")
table(seqnames(DHSA549Hg38))
```

Now e apply a segmented block bootstrap with blocks of size 500kb,
to the peaks. We are currently interested in profiling the speed of
this operation (future vignette iterations will also demonstrate its
use in a workflow).

Testing the segmented block bootstrap:

```{r}
blockLength <- 5e5
# TODO 'x' has no seqlengths
x <- DHSA549Hg38
seqlengths(x) <- seqlengths(seg)
```

```{r}
library(microbenchmark)
microbenchmark(
  list=alist(
    prop = bootRanges(x, seg, blockLength, 1, deny, proportionLength = TRUE),
    no_prop = bootRanges(x, seg, blockLength, 1, deny, proportionLength = FALSE)
), times=10)
```

# Visualizing the segmented block bootstrap

Below we present a toy example for visualizing the segmented block
bootstrap.

```{r}
suppressPackageStartupMessages(library(BentoBox))
my_palette <- function(n) {
  head(c("red","green3","red3","dodgerblue",
         "blue2","green4","darkred"), n)
}
plotGRanges <- function(gr) {
  bb_pageCreate(width = 5, height = 5, xgrid = 0,
                ygrid = 0, showGuides = TRUE)
  for (i in seq_along(seqlevels(gr))) {
    chrom <- seqlevels(gr)[i]
    chromend <- seqlengths(gr)[[chrom]]
    suppressMessages({
      p <- bb_params(chromstart = 0, chromend = chromend,
                     x = 0.5, width = 4*chromend/500, height = 2,
                     at = seq(0, chromend, 50),
                     fill = colorby("state_col", palette=my_palette))
      prngs <- bb_plotRanges(data = gr, params = p,
                             chrom = chrom,
                             y = 2 * i,
                             just = c("left", "bottom"))
      bb_annoGenomeLabel(plot = prngs, params = p, y = 0.1 + 2 * i)
    })
  }
}
```

```{r}
library(GenomicRanges)
seq_nms <- rep(c("chr1","chr2"), c(4,3))
seg <- GRanges(
  seqnames = seq_nms,
  IRanges(start = c(1, 101, 201, 401, 1, 201, 301),
          width = c(100, 100, 200, 100, 200, 100, 100)),
  seqlengths=c(chr1=500,chr2=400),
  state = c(1,2,1,3,3,2,1),
  state_col = factor(1:7)
)
```

```{r toysegments}
plotGRanges(seg)
```

```{r toyranges}
set.seed(1)
n <- 200
gr <- GRanges(
  seqnames=sort(sample(c("chr1","chr2"), n, TRUE)),
  IRanges(start=round(runif(n, 1, 500-10+1)), width=10)
)
suppressWarnings({
  seqlengths(gr) <- seqlengths(seg)
})
gr <- gr[!(seqnames(gr) == "chr2" & end(gr) > 400)]
gr <- sort(gr)
idx <- findOverlaps(gr, seg, type="within", select="first")
gr <- gr[!is.na(idx)]
idx <- idx[!is.na(idx)]
gr$state <- seg$state[idx]
gr$state_col <- factor(seg$state_col[idx])
plotGRanges(gr)
```

## Not scaling blocks to segmentation length

```{r toy-no-prop}
set.seed(1)
gr_prime <- bootRanges(gr, seg, blockLength = 25, R = 1, 
                       proportionLength = FALSE)[[1]]
plotGRanges(gr_prime)
```

## Scaling blocks to segmentation length

```{r toy-prop}
set.seed(1)
gr_prime <- bootRanges(gr, seg, blockLength = 50, R = 1, 
                       proportionLength = TRUE)[[1]]
plotGRanges(gr_prime)
```

