---
title: "Segmented block bootstrap"
author: "Wancen Mu"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
output:
  rmarkdown::html_document:
    highlight: tango
    toc: true
    toc_float: true	
    fig_width: 5
    fig_height: 3
vignette: |
  %\VignetteIndexEntry{Segmented block bootstrap}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction to segmented block bootstrap

cite Bickel et al...

Options:

* Segment genome using feature density, e.g. gene density
* Use exiting segmentation (e.g. ChromHMM, etc.)

# Segmentation by gene density

First we obtain the Ensembl genes ...

```{r}
library(EnsDb.Hsapiens.v86)
edb <- EnsDb.Hsapiens.v86
filt <- AnnotationFilterList(GeneIdFilter("ENSG", "startsWith"))
g <- genes(edb, filter = filt)
```

```{r}
library(GenomeInfoDb)
g <- keepStandardChromosomes(g, pruning.mode = "coarse")
seqlevels(g, pruning.mode="coarse") <- setdiff(seqlevels(g), "MT")
# seqlevelsStyle(g) <- "UCSC" # host issues with NCBI
seqlevels(g) <- paste0("chr", seqlevels(g))
genome(g) <- "hg38"
g <- sortSeqlevels(g)
g <- sort(g)
table(seqnames(g))
```

## Import deny region

Deny regions ... cite Kundaje

```{r}
library(nullrangesData)
data("deny")
table(seqnames(deny))
```

```{r}
library(nullranges)
library(plyranges)
```

## CBS segmentation

We subset the deny ranges to those which are 500 bp or larger.

cite: DNAcopy paper

```{r seg-cbs}
deny2 <- deny %>% plyranges::filter(width(deny) >= 500)
L_s <- 1e6
out <- segmentDensity(g, n = 3, L_s = L_s,
                       deny = deny2, type = "cbs", returnPlot = TRUE)
seg <- out$seg
print(out$p1)
print(out$p2)
```

## Alternatively: HMM segmentation

cite: RcppHMM

```{r seg-hmm}
out_hmm <- segmentDensity(g, n = 3, L_s = L_s,
                           deny = deny2, type = "hmm",returnPlot = TRUE)
print(out_hmm$p1)
print(out_hmm$p2)
```

# Segmented block bootstrap within chromosome

# Timing on DHS peaks

*DHSA549Hg38* First download DNase hypersensitivity peaks in A549 from
*AnnotationHub*, then subset to standard chromosomes and remove mitochondria (more work
needed to handle these cases internally). We use a "chain file" from UCSC to move the ranges from hg19 to the modern hg38 to be consistent with gene ranges. Finally, we sort the
features to be bootstrapped.

```{r}
data("DHSA549Hg38")
table(seqnames(DHSA549Hg38))
```

Here we apply a segmented block bootstrap with blocks of size 500kb,
to the peaks...

Test on 10 times bootstrap ~13s

```{r}
L_b <- 5e5
x <- DHSA549Hg38
system.time({
  res_wochr <- segBootstrapRanges(x, seg, L_b, R = 10,
                                  deny = deny, proportion_length = FALSE)
})
system.time({
  res_wochr <- segBootstrapRanges(x, seg, L_b, R = 10, 
                                  deny = deny, deny_option = "trim",
                                  proportion_length = FALSE)
})
## system.time({
##   res_wochr_propl <- segBootstrapRanges(x, seg, L_b, R = 10, deny = deny)
## })
system.time({
  res_wchr <- segBootstrapRanges(x, seg, L_b, R = 10, deny = deny,
                                 within_chrom = TRUE, proportion_length = FALSE)
})
## system.time({
##   res_wchr <- segBootstrapRanges(x, seg, L_b, R = 10, deny = deny,
##                                  within_chrom = TRUE, proportion_length = TRUE)
## })
```

# Visualizing the segmented block bootstrap

Toy example for evaluating segmented block bootstrap...

```{r}
suppressPackageStartupMessages(library(BentoBox))
plotGRanges <- function(gr) {
  bb_pageCreate(width = 5, height = 2, xgrid = 0,
                ygrid = 0, showGuides = FALSE)
  for (i in seq_along(seqlevels(gr))) {
    chrom <- seqlevels(gr)[i]
    chromend <- seqlengths(gr)[[chrom]]
    suppressMessages({
      p <- bb_params(chromstart = 0, chromend = chromend,
                     x = 0.5, width = 4*chromend/500, height = 0.6,
                     at = seq(0, chromend, 50),
                     fill = cols)
      pbed <- bb_plotBed(data = gr, params = p,
                         chrom = chrom,
                         y = 0.6 + (i-1),
                         just = c("left", "bottom"),
                         colorby=colorby("state_col"))
      bb_annoGenomeLabel(plot = pbed, params = p, y = 0.7 + (i-1))
    })
  }
}
```

```{r}
library(GenomicRanges)
seq_nms <- rep(c("chr1","chr2"), c(4,3))
seg <- GRanges(
  seqnames = seq_nms,
  IRanges(start = c(1, 101, 201, 401, 1, 201, 301),
          width = c(100, 100, 200, 100, 200, 100, 100)),
  seqlengths=c(chr1=500,chr2=400),
  state = c(1,2,1,3,3,2,1),
  state_col = c(1,2,4,3,6,5,7)
)
```

```{r toysegments}
cols <- c("red","green3","dodgerblue",
          "red3","green4","blue2",
          "darkred")
plotGRanges(seg)
```

```{r toyranges}
n <- 50
gr <- GRanges(
  seqnames=sort(sample(c("chr1","chr2"), n, TRUE)),
  IRanges(start=round(runif(n, 1, 500-20+1)), width=20)
)
suppressWarnings({
  seqlengths(gr) <- seqlengths(seg)
})
gr <- gr[!(seqnames(gr) == "chr2" & end(gr) > 400)]
gr <- sort(gr)
idx <- findOverlaps(gr, seg, select="first")
gr$state <- seg$state[idx]
gr$state_col <- seg$state_col[idx]
plotGRanges(gr)
```

```{r}
deny <- GRanges("chr1", IRanges(21,40), seqlengths=seqlengths(gr))
# why warning?
suppressWarnings({ 
  gr_prime <- nullranges::segBootstrapRanges(gr, seg, L_b=100, R=1, deny=deny)[[1]]
})
# plotGRanges(gr_prime)
```
