---
title: "Un-segmented block bootstrap"
author: "Michael Love"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
output:
  rmarkdown::html_document:
    highlight: tango
    toc: true
    toc_float: true
vignette: |
  %\VignetteIndexEntry{Toy data bootstrap ranges}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy=FALSE, cache=FALSE)
```

In this vignette, we demonstrate the un-segmented block bootstrap
functionality implemented in *nullranges*. "Un-segmented" refers to
the fact that this implementation does not consider segmentation of
the genome for sampling of blocks, see the segmented block bootstrap
vignette for the alternative implementation.

# Timing on DHS peaks

First download DNase hypersensitivity peaks in A549 from
*AnnotationHub*.
We subset to standard chromosomes and remove mitochondria (more work
needed to handle these cases internally). Finally, we sort the
features to be bootstrapped.

```{r eval=FALSE}
suppressPackageStartupMessages(library(AnnotationHub))
ah <- AnnotationHub()
pks <- query(ah, c("DnaseSeq", "narrowPeak", "A549", "UniPk"))[[1]]
library(GenomeInfoDb)
pks <- keepStandardChromosomes(pks)
seqlevels(pks) <- setdiff(seqlevels(pks), "chrM")
table(seqnames(pks))
pks <- sort(pks)
```

```{r results="hide"}
library(nullrangesData)
data("DHSA549Hg38")
pks <- DHSA549Hg38
# to get seqlengths only
suppressPackageStartupMessages(library(EnsDb.Hsapiens.v86))
edb <- EnsDb.Hsapiens.v86
filt <- AnnotationFilterList(GeneIdFilter("ENSG", "startsWith"))
g <- genes(edb, filter = filt)
g <- keepStandardChromosomes(g, pruning.mode = "coarse")
seqlevels(g, pruning.mode="coarse") <- setdiff(seqlevels(g), "MT")
seqlevels(g) <- paste0("chr", seqlevels(g))
genome(g) <- "hg38"
g <- sortSeqlevels(g)
seqlengths(pks) <- seqlengths(g)
```

```{r}
library(nullranges)
```

The following evaluates various types of bootstrap/permutation
schemes, first within chromosome, and then across chromosome (the
default). The default `type` is bootstrap.

```{r}
set.seed(5) # reproducibility
library(microbenchmark)
L_b <- 5e5
microbenchmark(list=alist(
                 p_within=bootRanges(pks, L_b=L_b, within_chrom=TRUE, type="permute"),
                 b_within=bootRanges(pks, L_b=L_b, within_chrom=TRUE),
                 p_across=bootRanges(pks, L_b=L_b, within_chrom=FALSE, type="permute"),
                 b_across=bootRanges(pks, L_b=L_b, within_chrom=FALSE)
), times=10)
```

# Visualize on synthetic data

We create some synthetic ranges in order to visualize the different
options of un-segmented bootstrap.
`chrom` will be used to color by chromosome of origin.

```{r}
library(GenomicRanges)
seq_nms <- rep(c("chr1","chr2","chr3"),c(4,5,2))
gr <- GRanges(seqnames=seq_nms,
              IRanges(start=c(1,101,121,201,
                              101,201,216,231,401,
                              1,101),
                      width=c(20, 5, 5, 30,
                              20, 5, 5, 5, 30,
                              80, 40)),
              seqlengths=c(chr1=300,chr2=450,chr3=200),
              chrom=as.integer(factor(seq_nms)))
```

The following function uses functionality from *BentoBox* to plot the ranges.

```{r}
suppressPackageStartupMessages(library(BentoBox))
plotGRanges <- function(gr) {
  bb_pageCreate(width = 5, height = 2, xgrid = 0,
                ygrid = 0, showGuides = FALSE)
  for (i in seq_along(seqlevels(gr))) {
    chrom <- seqlevels(gr)[i]
    chromend <- seqlengths(gr)[[chrom]]
    suppressMessages({
      p <- bb_params(chromstart = 0, chromend = chromend,
                     x = 0.5, width = 4*chromend/500, height = 0.5,
                     at = seq(0, chromend, 50),
                     fill = palette()[2:4])
      prngs <- bb_plotRanges(data = gr, params = p,
                             chrom = chrom,
                             y = 0.25 + (i-1)*.7,
                             just = c("left", "bottom"),
                             colorby=colorby("chrom"))
      bb_annoGenomeLabel(plot = prngs, params = p, y = 0.30 + (i-1)*.7)
    })
  }
}
```

```{r toyranges, fig.width=5, fig.height=2}
plotGRanges(gr)
```

## Within chromosome

Visualize two permutations of blocks within chromosome:

```{r perm-within, fig.width=5, fig.height=2}
for (i in 1:2) {
  gr_prime <- bootRanges(gr, L_b=100, type="permute", within_chrom=TRUE)[[1]]
  plotGRanges(gr_prime)
}
```

Visualize two bootstraps within chromosome:

```{r boot-within, fig.width=5, fig.height=2}
for (i in 1:2) {
  gr_prime <- bootRanges(gr, L_b=100, within_chrom=TRUE)[[1]]
  plotGRanges(gr_prime)
}
```

## Across chromosome

Visualize two permutations of blocks across chromosome. Here we use
larger blocks than previously.

```{r perm-across, fig.width=5, fig.height=2}
for (i in 1:2) {
  gr_prime <- bootRanges(gr, L_b=200, type="permute", within_chrom=FALSE)[[1]]
  plotGRanges(gr_prime)
}
```

Visualize two bootstraps across chromosome:

```{r boot-across, fig.width=5, fig.height=2}
for (i in 1:2) {
  gr_prime <- bootRanges(gr, L_b=200, within_chrom=FALSE)[[1]]
  plotGRanges(gr_prime)
}
```
